src/libraries/System.Reflection.Emit/src/System/Reflection/Emit/ModuleBuilderImpl.cs

```diff
@@ -119,6 +120,8 @@ internal void AppendMetadata()
 
             WriteCustomAttributes(_customAttributes, moduleHandle);
 
+            // All generic parameters for all types and methods should be written in specific order
+            ImmutableSortedDictionary<int, Type[]>.Builder paramsBuilder = ImmutableSortedDictionary.CreateBuilder<int, Type[]>();
```

Comment:
`ImmutableSortedDictionary` is fairly expensive datastructure. We do not need this to be immutable and sorted all the time. This can be `List<GenericTypeParameterBuilderImpl>` that is sorted right before we write it out.

Reply:
> ImmutableSortedDictionary is fairly expensive datastructure. We do not need this to be immutable and sorted all the time.

I assumed using ImmutableSortedDictionary<int, Type[]>.Builder would not be as expensive as `ImmutableSortedDictionary`. I assumed it would sort and create the `ImmutableSortedDictionary` only when `Builder.ToImmutableSortedDictionary()` called

> - This can be List<GenericTypeParameterBuilderImpl> that is sorted right before we write it out.

I tried that initially but the sorting APIs available are not in place and changing the order of the parameter having same parent. Well I could write the sorting manually and use it. But it seems to me using the `ImmutableSortedDictionary<int, Type[]>.Builder` is better because we don't need to sort params having same parent, only index will be sorted. We save the parameters Type[] array as is, no need to copy to another list and sort again

Reply:
> I assumed it would sort and create the ImmutableSortedDictionary only when Builder.ToImmutableSortedDictionary() called

Yes, it does - the AVL tree data structure that it creates is designed to allow further incremental updates. It is multiple times more expansive than sorting an array or list, that's all what we need here.

Reply:
What you think about of using a normal dictionary and sort the keys later before writing the parameters? I feel like it's better to keep the parameter Type[] as is as dictionary values ...

Reply:
The typical number of generic arguments is 1. It is fine to copy all generic arguments over to keep things simpler, we are not saving anything meaningful by keeping track of `Type[]`.

Also, it does not need to be a dictionary. It can be just a List.